client/asset/eth/eth.go

```diff
@@ -668,12 +619,166 @@ func (eth *ExchangeWallet) unlockFunds(coins asset.Coins) error {
+func (eth *ExchangeWallet) Swap(swaps *asset.Swaps) ([]asset.Receipt, asset.Coin, uint64, error) {
+	fail := func(err error) ([]asset.Receipt, asset.Coin, uint64, error) {
+		return nil, nil, 0, err
+	}
+
+	initiations := make([]dexeth.ETHSwapInitiation, 0, len(swaps.Contracts))
+	receipts := make([]asset.Receipt, 0, len(swaps.Contracts))
+
+	eth.lockedFundsMtx.Lock()
+	defer eth.lockedFundsMtx.Unlock()
+
+	var totalInputValue uint64
+	for _, input := range swaps.Inputs {
+		if _, ok := eth.lockedFunds[input.ID().String()]; !ok {
+			return fail(fmt.Errorf("Swap: attempting to use coin that was not locked"))
+		}
+		totalInputValue += input.Value()
+	}
+
+	var totalContractValue uint64
+	secretHashMap := make(map[string]bool)
+	for _, contract := range swaps.Contracts {
+		totalContractValue += contract.Value
+
+		if len(contract.SecretHash) != srveth.SecretHashSize {
+			return fail(
+				fmt.Errorf("Swap: expected secret hash of length %v, but got %v",
+					srveth.SecretHashSize, len(contract.SecretHash)))
+		}
+		encSecretHash := hex.EncodeToString(contract.SecretHash)
+		if _, ok := secretHashMap[encSecretHash]; ok {
+			return fail(fmt.Errorf("Swap: secret hashes must be unique"))
+		}
+		secretHashMap[encSecretHash] = true
+		var secretHash [32]byte
+		copy(secretHash[:], contract.SecretHash)
+
+		if !common.IsHexAddress(contract.Address) {
+			return fail(fmt.Errorf("Swap: address in contract is not valid %v", contract.Address))
+		}
+		address := common.HexToAddress(contract.Address)
+
+		initiations = append(initiations,
+			dexeth.ETHSwapInitiation{
+				RefundTimestamp: big.NewInt(0).SetUint64(contract.LockTime),
+				SecretHash:      secretHash,
+				Participant:     address,
+				Value:           big.NewInt(0).SetUint64(contract.Value),
+			})
+	}
+
+	// If the (baseFee + GasTipCap) in the block that the transaction is mined
+	// is less than the GasFeeCap, then some money will be saved. These funds
+	// will be unlocked when the entire order has finished executing.
+	maxPossibleFee := uint64(len(initiations)) * srveth.InitGas * swaps.FeeRate
+
+	totalUsedValue := totalContractValue + maxPossibleFee
+	if totalInputValue < totalUsedValue {
+		return fail(fmt.Errorf("Swap: coin inputs value %d < required %d", totalInputValue, totalUsedValue))
+	}
+
+	suggestedGasTipCap, err := eth.node.suggestGasTipCap(eth.ctx)
+	if err != nil {
+		return fail(fmt.Errorf("Swap: failed to get suggested gas tip cap %w", err))
+	}
+
+	var gasTipCap *big.Int
+	suggestedGasTipCapGwei, err := srveth.ToGwei(suggestedGasTipCap)
+	if err != nil {
+		return fail(fmt.Errorf("Swap: failed to convert to gwei: %w", err))
+	}
+	if suggestedGasTipCapGwei > MinGasTipCap {
+		gasTipCap = suggestedGasTipCap
+	} else {
+		gasTipCap = big.NewInt(MinGasTipCap * srveth.GweiFactor)
+	}
+
+	opts := bind.TransactOpts{
+		From:      eth.acct.Address,
+		Value:     big.NewInt(0).SetUint64(totalContractValue),
+		GasFeeCap: big.NewInt(0).SetUint64(swaps.FeeRate),
```

Comment:
Don't these also need to be converted from gwei to wei? I could be wrong, just let me know.

Reply:
Yes